### PR TITLE
[SLM] Integration with Disco sharding.

### PR DESCRIPTION
**Pull request by LeshengJin**

### CODE REVIEW

File: python/mlc_chat/compiler/compile.py [hunk 1]
```diff
@@ -94,6 +98,32 @@ def _attach_variable_bounds(mod, model_config):
...
+                continue
+            shard_info_dict[name].append((shard_func_name, shard_func_info))
+            if shard_func_name not in shard_funcs:
+                shard_funcs[shard_func_name] = shard_func
+        return shard_info_dict, shard_funcs
+
+    shard_info_dict, shard_funcs = _get_shard_info()
+    for name, func in shard_funcs.items():
+        func = func.with_attr({"global_symbol": name})
+        mod[name] = func
```
Comment (junrushao) [inline python/mlc_chat/compiler/compile.py:122]: nit: 

```suggestion
        mod[name] = func.with_attr({"global_symbol": name})
```

File: python/mlc_chat/distributed/worker.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+from tvm._ffi import get_global_func, register_func
+from tvm.runtime import NDArray, ShapeTuple, String
+from tvm.runtime.ndarray import array
+
+from ..base import _load_mlc_llm_lib
+
+
+@register_func("tests.disco.add_one")
+def _add_one(x: int) -> int:  # pylint: disable=invalid-name
+    return x + 1
```
Comment (junrushao) [inline python/mlc_chat/distributed/worker.py:32]: you probably don'y need those methods - for testing only

File: python/mlc_chat/distributed/worker.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+"""Internal DiscoWorker for Disco ProcessSession."""
+import os
+import sys
+
+from tvm import runtime as _  # pylint: disable=unused-import
+from tvm._ffi import get_global_func, register_func
+from tvm.runtime import NDArray, ShapeTuple, String
+from tvm.runtime.ndarray import array
+
+from ..base import _load_mlc_llm_lib
```
Comment (junrushao) [inline python/mlc_chat/distributed/worker.py:27]:
```suggestion
from .. import base
```

File: cpp/model_metadata.cc [hunk 4]
```diff
@@ -8,13 +8,45 @@
 namespace mlc {
 namespace llm {
 
+template <typename ExpectedType>
```
Comment (junrushao) [inline cpp/model_metadata.cc:11]: remove this

File: cpp/model_metadata.cc [hunk 5]
```diff
@@ -8,13 +8,45 @@
...
+ExpectedType AsType(const picojson::value& json) {
+  ICHECK(json.is<ExpectedType>());
+  return json.get<ExpectedType>();
+}
```
Comment (junrushao) [inline cpp/model_metadata.cc:15]: Move to `json_parser.h` as:

```C++
template <typename ValueType>
inline ValueType Lookup(const picojson::array& json, int index);
```

File: cpp/model_metadata.cc [hunk 6]
```diff
@@ -8,13 +8,45 @@
...
+  picojson::array shape_dtype = AsType<picojson::array>(json_preproc[1]);
+  CHECK_EQ(json_preproc.size(), 2) << "ValueError: Invalid preprocessing dtype shape in JSON";
+  picojson::array shape_json = AsType<picojson::array>(shape_dtype[0]);
+  int ndim = shape_json.size();
+  std::vector<int64_t> shape;
+  shape.reserve(ndim);
+  for (int i = 0; i < ndim; ++i) {
+    shape.push_back(AsType<int64_t>(shape_json[i]));
+  }
+  std::string dtype = AsType<std::string>(shape_dtype[1]);
```
Comment (junrushao) [inline cpp/model_metadata.cc:33]: Use existing `Lookup` methods

File: cpp/model_metadata.cc [hunk 7]
```diff
@@ -48,5 +80,27 @@ ModelMetadata ModelMetadata::FromModule(tvm::runtime::Module module) {
   }
 }
 
+std::string GetModelMetadataStringFromModule(tvm::runtime::Module module) {
```
Comment (junrushao) [inline cpp/model_metadata.cc:83]: Use this method in `ModelMetadata::FromModule`

File: cpp/loader/ndarray_cache_metadata.h [hunk 8]
```diff
@@ -0,0 +1,94 @@
+#ifndef MLC_LLM_CPP_LOADER_NDARRAY_CACHE_METADATA_H_
```
Comment (junrushao) [inline cpp/loader/ndarray_cache_metadata.h:1]: Remove this file, and instead in TVM Unity, move `ndarray_cache_metadata.h` to public API

File: cpp/loader/multi_gpu_loader.h [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+  NDArray LoadPresharded(int weight_index) const;
+
+  NDArray LoadFromModelMetaData(std::string param_name) const;
+
+  Array<NDArray> LoadAllFromModelMetaData() const;
+
+  /*! \brief Slice the given tensor at a specific dimension */
+  NDArray Shard(NDArray source, int dim, int num_slices) const;
+
+  static constexpr const char* _type_key = "runtime.disco.ShardLoader";
```
Comment (junrushao) [inline cpp/loader/multi_gpu_loader.h:59]: name change

File: python/mlc_chat/distributed/worker.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (jinhongyii) [inline python/mlc_chat/distributed/worker.py:1]: Why do we need a new worker class here?

File: python/mlc_chat/compiler/compile.py [hunk 11]
```diff
@@ -105,9 +134,12 @@ def _compile(args: CompileArgs):
     )
     logger.info("Running optimizations using TVM Unity")
     _attach_variable_bounds(mod, model_config)
+    shard_info_dict = _attach_shard_funcs(mod, named_params, model_config)
     with args.target:
         mod = relax.get_pipeline("mlc_llm")(mod)
-    _attach_auxiliary_methods(mod, named_params, args)
+    _attach_auxiliary_methods(mod, named_params, args, shard_info_dict)
+    with open("/opt/scratch/lesheng/mlc-llm/1.py", "w", encoding="utf-8") as outfile:
```
Comment (jinhongyii) [inline python/mlc_chat/compiler/compile.py:141]: remove logging

File: cpp/loader/multi_gpu_loader.cc [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+  f.CallPacked(TVMArgs(tvm_args.data(), type_codes.data(), n + 2), &rv);
+  return o;
+}
+
+std::unordered_map<std::string, NDArray> MultiGPULoaderObj::LoadFromRecord(int index) const {
+  std::unordered_map<std::string, NDArray> loaded_params;
+  DiscoWorker* worker = DiscoWorker::ThreadLocal();
+  Device device = worker->default_device;
+  const NDArrayCacheMetadata::FileRecord file_record = metadata_.records.at(index);
+  Array<String> loaded_param_names =
```
Comment (jinhongyii) [inline cpp/loader/multi_gpu_loader.cc:76]: we already have an implementation from https://github.com/apache/tvm/blob/unity/src/runtime/disco/loader.cc. Why do you re-implement a loading logic?

File: python/mlc_chat/compiler/help.py [hunk 13]
```diff
@@ -99,6 +99,9 @@
 (Experimental) The chunk size during prefilling. By default,
 the chunk size is the same as sliding window or max sequence length.
 This flag subjects to future refactoring.
+""".strip(),
+    """num_shards""": """
```
Comment (junrushao) [inline python/mlc_chat/compiler/help.py:103]: let's later rename it to something more informative, such as `num_tp_shards`, or `num_tp_workers`. i'm not sure

File: cpp/model_metadata.h [hunk 14]
```diff
@@ -2,6 +2,8 @@
  * \file model_metadata.h
  * \brief Metadata stored in model lib
  */
+#ifndef MLC_LLM_CPP_MODEL_METADATA_H_
```
Comment (junrushao) [inline cpp/model_metadata.h:5]: thanks for spotting this!

File: python/mlc_chat/compiler/help.py (hunk 13 above)
Comment (jinhongyii) [reply python/mlc_chat/compiler/help.py:103]: what about `tp_degree`?

File: cpp/json_parser.h [hunk 15]
```diff
@@ -21,23 +21,51 @@ inline ValueType Lookup(const picojson::object& json, const std::string& key) {
...
+template <typename ValueType>
+inline ValueType Lookup(const picojson::array& json, int index) {
+  CHECK(index < json.size()) << "IndexError: json::array index out of range";
+  auto value = json.at(index);
+  CHECK(value.is<ValueType>()) << "ValueError: value at index `" << index
+                               << "` has unexpected type";
+  return value.get<ValueType>();
+}
+
+tvm::runtime::DataType DataTypeFromString(const std::string& s) {
```
Comment (junrushao) [inline cpp/json_parser.h:33]: methods defined in the header have to be marked as "inline" to avoid ODR violations in most of the cases

**Comment by junrushao** [comment]
This PR is now functioning in:
* Model: Llama
* Quantization: q4f16_1
* num_shards: 2

Merging it now as it lays out a solid foundation, and we will need to quickly expand to:
* Models: Mistral, GPT2, GPTNeoX
* Quantization: AWQ and other group quantization
* num_shards: 4, 8, 16

**Comment by plasticchris** [comment]
Breaks sharding for me on:

model build
```
python3 -m mlc_llm.build --hf-path codellama/CodeLlama-34b-Python-hf \
    --target rocm \
    --quantization q4f16_1 \
    --max-seq-len 4096  --use-safetensors --use-cache=0 \
    --num-shards 2 --build-model-only
```
simple chat script
```
from mlc_chat import ChatModule, ChatConfig, ConvConfig
from mlc_chat.callback import StreamToStdout

cm = ChatModule(model="CodeLlama-34b-Python-hf-q4f16_1", chat_config=ChatConfig(tensor_parallel_shards=2))

while True:
    cm.generate(
        prompt=input(">"),
        progress_callback=StreamToStdout(callback_interval=2),
    )
    print(f"\nStatistics: {cm.stats()}\n")
```

Result: only 1 GPU in use. Reverted and using `num_shards=2` and it works again. I'm using 2x7900xtx and rocm.

**Pull request merged**